typed-lambda-calculus/article.md

```diff
@@ -0,0 +1,717 @@
+\end{array}
+$$
+
+$C$ represents type constructors, defined as primitives, and their set is arbitrary in HM, but it must include at least the abstraction type constructor $\to$.
+
+Parentheses can be used in value and type expressions to change evaluation order, but we omit these in the syntax definitions for brevity.
+
+Note that type variables are admitted as monotypes, which makes those distinct from _monomorphic types_ that can't have type variables.
+
+Polytypes also include monotypes, in addition to quantified types.
+
+The quantifier $\forall$ _binds_ the type variables, so in a type expression $\forall \alpha_1\;\ldots\;\alpha_n.\; \tau,$ all of $\alpha_i$ are _bound_.
+Type variables that are not bound are called _free_.
+
+Additionally, type variables may be bound by the typing context.
+The reason for this addition is nested let-bindings.
+
+We also need a function $free$ to define typing rules.
+$free$ collects all free variables in a type expression.
+It is defined as follows:
+
+$$
+\begin{array}{lcl}
+free(\alpha) & = & \{\alpha\} \\
+free(C\;\tau_1\;\ldots\;\tau_n) & = & \bigcup\limits_{i=1}^n free(\tau_i) \\
+free(\Gamma) & = & \bigcup\limits_{x:\sigma \in \Gamma} free(\sigma) \\
+free(\forall \alpha.\; \sigma) & = & free(\sigma) - \{\alpha\} \\
+free(\Gamma \vdash e:\sigma) & = & free(\sigma) - free(\Gamma) \\
+\end{array}
+$$
+
+If a type variable occurs unbound in a typing, it is implicitly treated as universally quantified.
+
+#### The principal type
+
+Polymorphism implies that the same term can have more than one type.
+In fact, it could have infinitely many types, as is the case with $\lambda x.\; x.$
+
+For the type system to be sound, we need to somehow choose the "most best" type out of all possibilities.
+Thankfully, there is a straightforward way to do so.
+
+The idea is to choose the largest most general type, also called the principal type -- that is, the most specific polytype that can still produce any other possible type via bound variable substitution.
+We denote substitution of a type variable $\alpha$ for a monotype $\tau$ in an expression $e$ as $\{\alpha \mapsto \tau\}e.$
+
+If some type $\tau'$ can be produced from $\tau$ by substitution, it is said that $\tau$ is more general than $\tau',$ written as $\tau \sqsubseteq \tau'.$
+$\sqsubseteq$ defines a partial order on types.
+Note that the more general type is "smaller" than the less general type.
+This is a feature of treating types as sets: the more general the type, the fewer possible values it can have.
+The most general type $\forall\alpha.\;\alpha$ is usually considered uninhabited (with a few caveats), i.e. it has no values at all.
+
+Formally, this all can be expressed by the following rule (called specialization rule):
+
+$$
+\begin{array}{c}
+\beta_i \notin free(\forall\alpha_1\ldots\alpha_n.\;\tau) \\
+\hline
+\forall\alpha_1\ldots\alpha_n.\;\tau
+\sqsubseteq \forall\beta_1\ldots\beta_m.\;\{\alpha_i\mapsto\tau_i\}\tau
```

Comment:
I was initially unsure how to read this rule, but then I realized that it assumes that `tau_i` draws from a set of types `tau_1 .. tau_n`, all of which can mention type variables from `beta_1 .. beta_m`. This is implicit in the rule and could be spelled out in an accompanying paragraph.